portal-backend/depmap/compound/legacy_utils.py

```diff
@@ -1,35 +1,88 @@
 import logging
 import pandas as pd
+import re
 import sqlalchemy as sa
+from sqlalchemy import nullslast, case  # type: ignore
+from typing import Optional
 
 from depmap.interactive import interactive_utils
+from depmap.database import db
 from depmap.dataset.models import (
     Compound,
     CompoundExperiment,
+    DependencyDataset,
 )
 from depmap.partials.matrix.models import Matrix, RowMatrixIndex
 
 # These methods can all be deleted once compound datasets are migrated to breadbox.
 
 log = logging.getLogger(__name__)
 
-def get_compound_labels_for_compound_experiment_dataset(dataset_name: str) -> dict[str, str]:
+def _get_deduplicated_experiment_compound_mapping(dataset_id: str) -> list[tuple[CompoundExperiment, Compound]]:
     """ 
-    This method loads a mapping between the old index (compound experiment label) and 
-    method is used to re-index datasets which are indexed by compound experiment. 
+    Returns a 1-1 mapping between CompoundExperiments and Compounds within the given dataset. 
+    All compound experiments not in the mapping should be dropped. 
     """
-    matrix_id = interactive_utils.get_matrix_id(dataset_name)
+    matrix_id = interactive_utils.get_matrix_id(dataset_id)
     comp_exp_alias = sa.orm.aliased(CompoundExperiment)
     compound_alias = sa.orm.aliased(Compound)
-    labels_by_indeces = (
+    comp_exp_info = (
         Matrix.query.filter_by(matrix_id=matrix_id)
         .join(RowMatrixIndex)
         .join(comp_exp_alias)
         .join(compound_alias, compound_alias.entity_id == comp_exp_alias.compound_id)
-        .with_entities(comp_exp_alias.label, compound_alias.label)
+        .with_entities(comp_exp_alias, compound_alias)
         .all()
     )
-    return {experiment_id: compound_label for experiment_id, compound_label in labels_by_indeces}
+    # Organize values by compound ID
+    experiments_by_compound_id: dict[str, list] = {}
+    compounds_by_compound_id: dict[str, Compound] = {}
+    for comp_exp, compound in comp_exp_info:
+        compound_id = compound.entity_id
+        if compound_id in experiments_by_compound_id:
+            # Add to the existing list of experiments
+            experiments_by_compound_id[compound_id].append(comp_exp)
+        else:
+            # Add records to both dictionaries
+            experiments_by_compound_id[compound_id] = [comp_exp]
+            compounds_by_compound_id[compound_id] = compound
+
+    # Remove duplicate compound experiments, re-index by compound experiment
+    result = []
+    for compound_id, compound in compounds_by_compound_id.items():
+        experiments = experiments_by_compound_id[compound_id]
+        selected_experiment = sorted(experiments, key=lambda e: e.entity_id)[0]
+        result.append((selected_experiment, compound))
+    return result
```

Comment:
Refactored this method to use Phil's suggestion from [here](https://github.com/broadinstitute/depmap-portal/pull/175#discussion_r1930630207) (of collapsing by compound earlier in the process).